replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/Batch.scala

```diff
@@ -13,140 +30,103 @@ private[journal] sealed abstract class Batch extends Product {
+        )
+      } else {
+        this
+      }
+    }
 
-        bs match {
-          case b :: tail =>
-            (b, record.action) match {
-              case (b: Appends, a: Action.Append) =>
-                val records = actionRecord(a) :: b.records
-                appendsOf(records) :: tail
-
-              case (b: Appends, _: Action.Mark) =>
-                appendsOf(b.records) :: tail
-
-              case (b: Appends, a: Action.Delete) =>
-                if (cut(b, a)) {
-                  val delete = deleteOf(a.to, origin orElse a.origin, version orElse a.version)
-                  // preserve last `append` as we want to update `expireAfter` too, but we do not have [easy] access to it
-                  delete :: b.copy(records = NonEmptyList.of(b.records.head)) :: Nil
-                } else {
-                  val delete = deleteOf(a.to, a.origin, a.version)
-                  delete :: bs
-                }
-
-              case (_: Appends, a: Action.Purge) =>
-                purgeOf(a.origin, a.version) :: Nil
-
-              case (b: Delete, a: Action.Append) =>
-                appendsOf(NonEmptyList.of(actionRecord(a))) :: b :: tail
-
-              case (b: Delete, _: Action.Mark) =>
-                b.copy(offset = offset) :: tail
-
-              case (b: Delete, a: Action.Delete) =>
-                if (a.to > b.to) {
-                  val cleanTail: List[Batch] = tail.mapFilter {
-                    case appends: Appends => dropDeleted(appends, a)
-                    case delete: Delete   => delete.some
-                    case purge: Purge     => purge.some
-                  }
-
-                  val delete = deleteOf(a.to, b.origin orElse a.origin, b.version orElse a.version)
-                  delete :: cleanTail
-                } else {
-                  val delete = b.copy(offset = offset, origin = b.origin orElse a.origin, version = b.version orElse a.version)
-                  delete :: tail
-                }
-
-              case (_: Delete, a: Action.Purge) =>
-                purgeOf(a.origin, a.version) :: Nil
-
-              case (b: Purge, a: Action.Append) =>
-                appendsOf(NonEmptyList.of(actionRecord(a))) :: b :: Nil
-
-              case (b: Purge, _: Action.Mark) =>
-                b.copy(offset = offset) :: Nil
-
-              case (b: Purge, a: Action.Delete) =>
-                deleteOf(a.to, a.origin, a.version) :: b :: Nil
-
-              case (_: Purge, a: Action.Purge) =>
-                purgeOf(a.origin, a.version) :: Nil
-            }
-
-          case Nil =>
-            record.action match {
-              case a: Action.Append => appendsOf(NonEmptyList.of(actionRecord(a))) :: Nil
-              case _: Action.Mark   => Nil
-              case a: Action.Delete => deleteOf(a.to, a.origin, a.version) :: Nil
-              case a: Action.Purge  => purgeOf(a.origin, a.version) :: Nil
-            }
+    def batches: List[Batch] = {
```

Comment:
The idea of creating at max three batches out of any number of events leads to splitting the logic on two parts: one for combining events into "batch"es (actually it should not be called batch anymore because its not just collapsed operations but rather instructions for underline logic, not directly connected with original events. As an example check `Batch.Delete` with offset coming from append) and another for hacking some values to fit preserved order (purge, appends, delete) into logic of persisting events in `ReplicatedCassandra`.

I think its not a way we should go